### PR TITLE
A couple of changes (X-REAL-IP & User Agent)

### PR DESCRIPTION
**Pull request by ajarmoszuk**
Here's the list of changes I've made:
- Use X-REAL-IP (or X-FORWARDED-FOR) only when going through a proxy in a private CIDR and then CLIENT_IP as a fallback
- Add User Agent as a part of challenge response authentication cookie
- Change authentication cookie from "res" to "challenge_token" to make it less ambigious
- Tidy up README.md
- Added a error message for when no cookies are present
- Created a Github Action to build the module

### CODE REVIEW

**Comment by bnchdan** [comment]
You can easily spoofed the address IP. If is not behind a proxy.
```
curl https://my_server
   -H 'X-REAL-IP: value'
```

 In my opinion a better approach is to have an array of proxies that can set the X-REAL-IP.

File: ngx_http_js_challenge.c [hunk 1]
```diff
@@ -335,32 +341,61 @@ int get_cookie(ngx_http_request_t *r, ngx_str_t *name, ngx_str_t *value) {
...
 
-    unsigned long bucket = r->start_sec - (r->start_sec % conf->bucket_duration);
-    ngx_str_t addr = r->connection->addr_text;
+    // Check for X-REAL-IP header and fallback to connection IP if not present
+    ngx_str_t addr = r->connection->addr_text; // Default IP
+    ngx_list_part_t *part = &r->headers_in.headers.part;
+    ngx_table_elt_t *header = part->elts;
+    for (ngx_uint_t i = 0; i < part->nelts; i++) {
+        if (ngx_strncasecmp(header[i].key.data, (u_char *)"X-REAL-IP", header[i].key.len) == 0 && header[i].value.len > 0) {
+            addr = header[i].value;  // Use X-REAL-IP if available
```
Comment (bnchdan) [inline ngx_http_js_challenge.c]: What will happen if the new X-REAL-IP have length greater than r->connection->addr_text ?

File: ngx_http_js_challenge.c [hunk 2]
```diff
@@ -335,32 +341,61 @@ int get_cookie(ngx_http_request_t *r, ngx_str_t *name, ngx_str_t *value) {
...
+            break;
+        }
+        if (i == part->nelts - 1 && part->next != NULL) {
+            part = part->next;
+            header = part->elts;
+            i = 0;
+        }
+    }
 
+    // Extract User-Agent header
```
Comment (bnchdan) [inline ngx_http_js_challenge.c:447]: you can access user agent from r->headers_in.user_agent struct

Comment (ajarmoszuk) [reply ngx_http_js_challenge.c:447]: I've cleaned up that code now.

File: ngx_http_js_challenge.c (hunk 1 above)
Comment (ajarmoszuk) [reply ngx_http_js_challenge.c]: Added a limit of 39 chars.

**Comment by ajarmoszuk** [comment]
> You can easily spoofed the address IP. If is not behind a proxy.
> 
> ```
> curl https://my_server
>    -H 'X-REAL-IP: value'
> ```
> 
> In my opinion a better approach is to have an array of proxies that can set the X-REAL-IP.

I've modified to only allow for proxies in the private CIDR range. While I didn't add a config for such I didn't think it would be required to be completely honest.

**Comment by bnchdan** [comment]
> > You can easily spoofed the address IP. If is not behind a proxy.
> > ```
> > curl https://my_server
> >    -H 'X-REAL-IP: value'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In my opinion a better approach is to have an array of proxies that can set the X-REAL-IP.
> 
> I've modified to only allow for proxies in the private CIDR range. While I didn't add a config for such I didn't think it would be required to be completely honest.

E.g. if you set "X-REAL-IP: 127.0.0.1"  in the header for requests from different IPs you will get the same challenge.

**Comment by ajarmoszuk** [comment]
Regrettably, this is the limit of the work I can undertake given my current circumstances. However, if anyone wishes to utilize my version, they are welcome to access it from my repository. I'm closing this pull request since it appears that this repository author is not currently active.

**Pull request closed without merging**